src/pybind/mgr/object_format.py

```diff
@@ -0,0 +1,487 @@
+    """A return-value adapter for an object.
+    Given an input object, this type will attempt to get a mgr return value
+    from the object if provides a `mgr_return_value` function.
+    If not it returns a default return value, typically 0.
+    """
+
+    def __init__(
+        self,
+        obj: Any,
+        default: int = 0,
+    ) -> None:
+        self.obj = obj
+        self.default_return_value = default
+
+    def mgr_return_value(self) -> int:
+        if isinstance(self.obj, ReturnValueProvider):
+            return int(self.obj.mgr_return_value())
+        return self.default_return_value
+
+
+class ErrorResponseBase(Exception):
+    """An exception that can directly be converted to a mgr reponse."""
+
+    def format_response(self) -> Tuple[int, str, str]:
+        raise NotImplementedError()
+
+
+class UnknownFormat(ErrorResponseBase):
+    """Raised if the format name is unexpected.
+    This can help distinguish typos from formats that are known but
+    not implemented.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unknown format name: {self.format_name}"
+
+
+class UnsupportedFormat(ErrorResponseBase):
+    """Raised if the format name does not correspond to any valid
+    conversion functions.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unsupported format: {self.format_name}"
+
+
+class ErrorResponse(ErrorResponseBase):
+    """General exception convertible to a mgr response."""
+
+    E = TypeVar("E", bound="ErrorResponse")
+
+    def __init__(self, status: str, return_value: Optional[int] = None) -> None:
+        self.return_value = (
+            return_value if return_value is not None else -errno.EINVAL
+        )
+        self.status = status
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return (self.return_value, "", self.status)
+
+    def mgr_return_value(self) -> int:
+        return self.return_value
+
+    @property
+    def errno(self) -> int:
+        rv = self.return_value
+        return -rv if rv < 0 else rv
+
+    def __repr__(self) -> str:
+        return f"ErrorResponse({self.status!r}, {self.return_value!r})"
+
+    @classmethod
+    def wrap(
+        cls: Type[E], exc: Exception, return_value: Optional[int] = None
+    ) -> E:
+        if return_value is None:
+            try:
+                return_value = -int(getattr(exc, "errno"))
+            except (AttributeError, ValueError):
+                pass
+        err = cls(str(exc), return_value=return_value)
+        setattr(err, "__cause__", exc)
+        return err
+
+
+def _get_requested_format(f: ObjectResponseFuncType, kw: Dict[str, Any]) -> str:
+    # todo: leave 'format' in kw dict iff its part of f's signature
```

Comment:
typo: iff

Reply:
Ah, this is not a typo, it's shorthand for if-and-only-if.  (https://en.wikipedia.org/wiki/If_and_only_if) I'm probably abusing the real mathematical term's meaning but it's spelling was intentional. :-)